File: src/lib/smoother.ts
import type { GestureName, GestureResult } from '../types';

// ─────────────────────────────────────────────
// Rolling-buffer gesture smoother
// Prevents jitter by only confirming a gesture
// when ALL frames in the buffer agree.
// ─────────────────────────────────────────────

export class GestureSmoother {
  private buffer: GestureName[] = [];
  private _bufferSize: number;

  constructor(bufferSize: number = 6) {
    this._bufferSize = bufferSize;
  }

  /** Update the buffer size (e.g. from settings slider). Clears existing buffer. */
  setBufferSize(size: number): void {
    this._bufferSize = size;
    this.buffer = [];
  }

  get bufferSize(): number {
    return this._bufferSize;
  }

  /**
   * Push a new raw gesture into the rolling buffer and return the
   * current smoothed result.
   */
  push(gesture: GestureName): GestureResult {
    this.buffer.push(gesture);
    if (this.buffer.length > this._bufferSize) {
      this.buffer.shift();
    }

    // Count how many frames match the most-recent gesture
    const candidate = this.buffer[this.buffer.length - 1];
    const confidence = this.buffer.filter((g) => g === candidate).length;
    const confirmed = this.buffer.length === this._bufferSize && confidence === this._bufferSize;

    return {
      gesture: candidate,
      confidence,
      confirmed,
      bufferSize: this._bufferSize,
    };
  }

  /** Reset the buffer */
  reset(): void {
    this.buffer = [];
  }
}

File: gesture_control/smoother.py
class GestureSmoother:
    def __init__(self, buffer_size=5):
        self.buffer_size = buffer_size
        self.buffer = []

    def push(self, gesture):
        """
        Pushes a new gesture into the rolling buffer.
        Returns the gesture only if all frames in the buffer agree,
        otherwise returns "NONE".
        """
        self.buffer.append(gesture)
        if len(self.buffer) > self.buffer_size:
            self.buffer.pop(0)

        # Confirm only if buffer is full and all gestures match
        if len(self.buffer) == self.buffer_size and all(g == self.buffer[-1] for g in self.buffer):
            return self.buffer[-1]
            
        return "NONE"

    def reset(self):
        """Clears the buffer."""
        self.buffer = []
